app/src/main/java/com/asfoundation/wallet/ui/onboarding/OnboardingPresenter.kt

```diff
@@ -1,74 +1,138 @@
+            .observeOn(viewScheduler)
+            .subscribe({ onNext(it, false) }, {})
+    )
   }
 
-  private fun handleLinkClick() {
-    view.getLinkClick()
-        ?.doOnNext { uri ->
-          view.navigateToBrowser(Uri.parse(uri))
-        }
-        ?.subscribe()
-        ?.let { disposables.add(it) }
+  private fun handleLaterClicks() {
+    disposables.add(
+        view.getLaterButtonClicks()
+            .doOnNext {
+              finishOnBoarding(null, false)
+            }.subscribe())
+  }
+
+  private fun handleRedeemButtonClicks() {
+    disposables.add(
+        Observable.zip(isWalletCreated(), view.getRedeemButtonClick(),
+            BiFunction { _: Boolean, _: Any -> }
+        )
+            .flatMapSingle { onboardingInteract.getWalletAddress() }
+            .flatMapSingle {
+              smsValidationInteract.isValid(Wallet(it))
+                  .subscribeOn(networkScheduler)
+            }
+            .delay(1, TimeUnit.SECONDS)
+            .observeOn(viewScheduler)
+            .subscribe({ onNext(it, true) }, {})
+    )
   }
 
-  private fun handleOnBoardingFinish() {
+  private fun onNext(walletValidationStatus: WalletValidationStatus?, showAnimation: Boolean) {
+    if (walletValidationStatus == WalletValidationStatus.NO_NETWORK) {
+      view.showNoInternetView()
+    } else {
+      finishOnBoarding(walletValidationStatus, showAnimation)
+    }
+  }
+
+  private fun handleNextButtonClicks() {
     disposables.add(
-        Observable.zip(handleGetWalletAddress(), handleSkip(),
-            BiFunction { _: String, _: Any -> true })
+        Observable.zip(isWalletCreated(), view.getNextButtonClick(),
+            BiFunction { _: Boolean, _: Any -> }
+        )
             .delay(1, TimeUnit.SECONDS)
             .observeOn(viewScheduler)
-            .subscribe({ finishOnBoarding() }, {})
+            .subscribe({ finishOnBoarding(null, true) }, {})
     )
   }
 
-  private fun handleGetWalletAddress(): Observable<String> {
-    return onboardingInteract.getWalletAddress()
-        .onErrorResumeNext {
-          onboardingInteract.createWallet()
+  private fun handleCreateWallet() {
+    disposables.add(
+        onboardingInteract.getWalletAddress()
+            .onErrorResumeNext {
+              onboardingInteract.createWallet()
+            }
+            .flatMapCompletable { Completable.fromAction { walletCreated.onNext(true) } }
+            .subscribe())
+  }
+
+  private fun handleSkipedOnboarding() {
+    disposables.add(
+        Observable.zip(isWalletCreated(),
+            Observable.fromCallable { onboardingInteract.hasClickedSkipOnboarding() }.filter { clicked -> clicked },
+            BiFunction { _: Boolean, _: Any -> }
+        )
+            .delay(1, TimeUnit.SECONDS)
+            .observeOn(viewScheduler)
+            .subscribe({ finishOnBoarding(null, true) }, {})
+    )
+  }
+
+  private fun handleLinkClick() {
+    view.getLinkClick()
+        ?.doOnNext { uri ->
```

Comment:
Not yours but can you remove the nullability from the getLinkMethod and add “!!” to the subject